feat(ivy): implement UMD for ngcc

petebacondarwin opened the pull request:
Add support to ngcc for UMD formatted packages.

Sits on top of #29643

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/fesm2015_host.ts:
```diff
@@ -241,13 +241,19 @@ export class Fesm2015ReflectionHost extends TypeScriptReflectionHost implements
+            if (ts.isPropertyAccessExpression(typeIdentifier)) {
+              // the type is in a namespace, e.g. core. Directive
+              typeIdentifier = typeIdentifier.name;
+            }
```
In UMD the type of a decorator can be a property access, e.g. `core.Directive` rather than just `Directive`.

petebacondarwin commented on packages/compiler-cli/src/ngcc/test/helpers/utils.ts:
```diff
@@ -14,6 +14,10 @@ export function makeProgram(...files: {name: string, contents: string}[]): ts.Pr
+export function makeProgramAndHost(...files: {name: string, contents: string}[]): {program: ts.Program, host: ts.CompilerHost} {
+  return _makeProgram([getFakeCore(), ...files], {allowJs: true, checkJs: false});
+}
+
```
The UMD reflection host needs access to the program and the compiler host.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/testing/in_memory_typescript.ts:
```diff
@@ -112,33 +112,37 @@ export function getDeclaration<T extends ts.Declaration>(
-  });
-
-  chosenDecl = chosenDecl as ts.Declaration | null;
+  const chosenDecl = walkForDeclaration(sf);
```
Previously we just looked at the top level statements for declarations but with UMD we need to traverse down into the file.

mary-poppins commented:
You can preview 215d9d2 at https://pr25445-215d9d2.ngbuilds.io/.

JoostK reviewed (commented):
Nice! 👏

mary-poppins commented:
You can preview 03b3a45 at https://pr25445-03b3a45.ngbuilds.io/.

JoostK commented on packages/compiler-cli/src/ngcc/src/rendering/umd_renderer.ts:
```diff
@@ -0,0 +1,51 @@
+import {Esm2015Renderer} from './esm2015_renderer';
+import {UmdReflectionHost, getUmdFactoryFunction, isRequireCall} from '../host/umd_host';
+
+export class UmdRenderer extends Esm2015Renderer {
```
This should extend `Esm5Renderer` for correctness, although it doesn't make a difference atm.

mary-poppins commented:
You can preview c96ca1f at https://pr25445-c96ca1f.ngbuilds.io/.

mary-poppins commented:
You can preview 96787b8 at https://pr25445-96787b8.ngbuilds.io/.

mary-poppins commented:
You can preview 24b7dd6 at https://pr25445-24b7dd6.ngbuilds.io/.

mary-poppins commented:
You can preview d25ad12 at https://pr25445-d25ad12.ngbuilds.io/.

mary-poppins commented:
You can preview 9684f96 at https://pr25445-9684f96.ngbuilds.io/.

mary-poppins commented:
You can preview b7d3e3a at https://pr25445-b7d3e3a.ngbuilds.io/.

mary-poppins commented:
You can preview e612945 at https://pr25445-e612945.ngbuilds.io/.

mary-poppins commented:
You can preview a6f7b09 at https://pr25445-a6f7b09.ngbuilds.io/.

petebacondarwin commented:
We need to account for the changes here: https://github.com/rollup/rollup/pull/2594

petebacondarwin committed:
test(ivy): ngcc - check the actual file that is passed to `renderImports`

Previously we were just checking that the object was "any" object but now
we check that it is the file object that we expected.

petebacondarwin committed:
refactor(ivy): use a named type for ImportManager import structures

Previously we were using an anonymous type `{specifier: string; qualifier: string;}`
throughout the code base. This commit gives this type a name and ensures it
is only defined in one place.

petebacondarwin committed:
refactor(ivy): ngcc - fake core and tslib should be typings files

Previously these fake files were full TypeScript source
files (`.ts`) but this is not necessary as we only need the
typings not the implementation.

petebacondarwin committed:
test(ivy): fix ESM5 test code to use `var` rather than `const`

petebacondarwin committed:
refactor(ivy): ngcc - abstract how module statements are found

This will be important for UMD support.

petebacondarwin committed:
feat(ivy): ngcc - implement `UmdReflectionHost`

petebacondarwin committed:
fix(ivy): ngcc - support namespaced identifiers

In UMD formats, imports are always namespaced. This commit makes
ngcc more tolerant of such structures.

petebacondarwin committed:
feat(ivy): ngtsc - support namespaced `forwardRef` calls

In some cases the `forwardRef` helper has been imported via a namespace,
e.g. `core.forwardRef(...)`.

This commit adds support for unwrapping such namespaced imports when
ngtsc is statically evaluating code.

petebacondarwin committed:
test(ivy): ngcc - remove unnecessary code

petebacondarwin committed:
test(ivy): enhance the in-memory-typescript helper

The `getDeclaration()` function now searches down into the AST for
matching nodes, which is needed for UMD testing.

petebacondarwin committed:
feat(ivy): ngcc - implement UmdDependencyHost

The dependency resolution that works out the order in which
to process entry-points must also understand UMD formats.

petebacondarwin committed:
feat(ivy): ngcc - implement `UmdRenderer`

petebacondarwin committed:
feat(ivy): ngcc - turn on UMD processing

petebacondarwin committed:
fix(ivy): ngcc - separate typings rendering from src rendering

Previously the same `Renderer` was used to render typings (.d.ts)
files. But the new `UmdRenderer` is not able to render typings files
correctly.

This commit splits out the typings rendering from the src rendering.
To achieve this the previous renderers have been refactored from
sub-classes of the abstract `Renderer` class to  classes that implement
the `RenderingFormatter` interface, which are then passed to the
`Renderer` and `DtsRenderer` to modify its rendering behaviour.

Along the way a few utility interfaces and classes have been moved
around and renamed for clarity.

alxhub commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -0,0 +1,268 @@
+    if (sourceFile.isDeclarationFile) {
+      return null;
+    }
+    let umdModule = this.umdModules.get(sourceFile);
```
Use `.has` instead of checking for `undefined`.

alxhub commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -0,0 +1,268 @@
+    let path = this.umdImportPaths.get(importParameter);
+    if (path === undefined) {
+      const umdModule = this.getUmdModule(importParameter.getSourceFile());
+      const imports = umdModule && getImportsOfUmdModule(umdModule);
```
Use `if` instead of `&&` short-circuiting.

alxhub commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -0,0 +1,268 @@
+      const umdModule = this.getUmdModule(importParameter.getSourceFile());
+      const imports = umdModule && getImportsOfUmdModule(umdModule);
+      if (imports !== null) {
+        imports.forEach(i => {
```
Prefer `for (const i of imports)` when iterating over arrays.

alxhub commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -0,0 +1,268 @@
+
+  /** Get the top level statements for a module.
+   *
+   * In ES5 and ES2015 this is just the top level statements of the file.
```
Comment is copy-pasted :)

petebacondarwin replied:
😱

alxhub commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -0,0 +1,268 @@
+    const name = statement.expression.left.name.text;
+
+    const declaration = this.getDeclarationOfExpression(exportExpression);
+    return declaration && {name, declaration};
```
Prefer ternary to `&&`.

In general I'm not a fan of the logical operators being used as control flow - I think it significantly affects readability to have multiple paths through the code without clear control flow operations.

petebacondarwin replied:
In this case I didn't really think of it as control flow. But happy to change.

alxhub commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -0,0 +1,268 @@
+
+  private getUmdImportedDeclaration(id: ts.Identifier): Declaration|null {
+    const importInfo = this.getImportOfIdentifier(id);
+    if (!importInfo) {
```
Prefer explicit checks for `=== null`.

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -1327,27 +1320,48 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+          return {...exportDecl, viaModule: namespaceDecl.viaModule};
+        }
+      }
+    }
```
The problem with putting this `return null;` in an else is that we then end up with a number of `undefined` return paths. But I guess this demonstrates how hidden those were. So I am going to flip the `if` checks to flatten the pyramid, which should make things clearer.

alxhub commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -0,0 +1,279 @@
+
+  getUmdExports(sourceFile: ts.SourceFile): Map<string, Declaration>|null {
+    let moduleExports: Map<string, Declaration>|null|undefined = this.umdExports.get(sourceFile);
+    if (moduleExports === undefined) {
```
Use `has` instead of checking for `undefined` to test presence/absence of a key in a `Map`.

alxhub commented on packages/compiler-cli/ngcc/src/rendering/dts_renderer.ts:
```diff
@@ -0,0 +1,211 @@
+        const dtsDeclaration = this.host.getDtsDeclaration(compiledClass.declaration);
+        if (dtsDeclaration) {
+          const dtsFile = dtsDeclaration.getSourceFile();
+          const renderInfo = dtsMap.get(dtsFile) || new DtsRenderInfo();
```
Use `has` with a ternary instead of the `||` operator as control flow.

alxhub commented on packages/compiler-cli/ngcc/src/rendering/dts_renderer.ts:
```diff
@@ -0,0 +1,211 @@
+        }
+      });
+      const dtsEntryPoint = this.bundle.dts !.file;
+      const renderInfo = dtsMap.get(dtsEntryPoint) || new DtsRenderInfo();
```
Same here.

alxhub commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -877,14 +874,20 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
-              import: this.getImportOfIdentifier(typeIdentifier), node,
-              args: getDecoratorArgs(node),
-            });
+          let typeIdentifier = decorator.get('type');
```
`.has`.

alxhub commented on packages/compiler-cli/ngcc/src/rendering/dts_renderer.ts:
```diff
@@ -0,0 +1,211 @@
+    // Capture the ModuleWithProviders functions/methods that need updating
+    if (moduleWithProvidersAnalyses !== null) {
+      moduleWithProvidersAnalyses.forEach((moduleWithProvidersToFix, dtsFile) => {
+        const renderInfo = dtsMap.get(dtsFile) || new DtsRenderInfo();
```
`.has` + ternary.

alxhub reviewed (approved):
👍 

A couple more points about `.has` and using `&&` and `||` as control flow.

It occurs to me that maybe I should write a TS style guide for the compiler, outlining the opinions I'm trying to enforce in the code, as well as the rationale for both of them.

petebacondarwin commented:
Agh! Sorry @alxhub I should have been more careful and fixed all of those.

petebacondarwin committed:
refactor(ivy): ngcc - use `.has()` to check Map membership

Previously we were relying upon the `.get()` method to return `undefined`
but it is clearer and safer to always check with `.has()` first.

petebacondarwin commented:
@jasonaden - conflicts resolved and rebased on master.
